### PR TITLE
Restore task.setAnnotation to tasks

### LINKED ISSUE #1951: Sub-Task Popup shows sub-task for most recent Drawing Mark, not currently selected Drawing Mark
## Major Issue

- Expected: when I click on a Drawing Mark (with a sub-task) to select it, the sub-task popup should appear, and the popup should show the sub-task for that **selected Drawing Mark.**
- Actual: the sub-task popup appears, it shows the sub-task for the **newest Drawing Mark**

### Testing

- Go to [blicksams's Transcription Task Testing](https://localhost:8080/?project=11300&env=production) (or any Workflow with Drawing Tasks that have sub-tasks, really)
- click to create your first Text Transcription Line 1️⃣ .
  - The sub-task popup should appear, with an empty text box.
  - for the transcription text, type in "Apple" 🍎.
  - Close the sub-task popup.
- click to create your second Text Transcription Line 2️⃣ .
  - The sub-task popup should appear, with an empty text box.
  - for the transcription text, type in "Banana" 🍌 .
  - Close the sub-task popup.
- click on the first Transcription Line 1️⃣ again.
  - The sub-task popup should appear with some text.
    - Expected: the sub-task should show the text typed in for the first line 1️⃣ , which is "Apple". 🍎 
    - ❌ Actual: the sub-task shows the text typed for the second line 2️⃣ , which is "Banana". 🍌 
  - Type in some new text, such as "Avocado" 🥑 
  - Close the sub-task popup.
- Submit the classification and monitor the submitted data.
  - Expected: the classification should submit annotation data that looks like `["Avocado", "Banana"]`
  - Actual: the classification submits annotation data that looks like `["Apple", "Avocado"]`

Tested on localhost with macOS + Chrome (on normal and with touchscreen emulation) on [blicksams's Transcription Task Testing](https://localhost:8080/?project=11300&env=production), which features a text transcription task.

**Advanced Testing:**

- On a new Subject, create the first Text Transcription Line 1️⃣ 
- Create the second Text Transcription Line 2️⃣ 
- Delete the second Text Transcription Line 2️⃣ 
- Click on the first Text Transcription Line 1️⃣ 
  - Expected: the sub-task popup should appear with the first line's sub-task. 1️⃣ 
  - Actual: classifier crashes because it's trying to access the sub-task of a deleted Mark.

### History

This issue was discovered while testing PR #1950 . After additional tests, I discovered this problem exists as far back as commit dc114e3, so this problem doesn't seem to have been introduced by the refactor in #1922 / #1921 (which 1950 was a response to).

### Status

n/a, to be updated

### PR DESCRIPTION
**Pull request by eatyourgreens**
Set a reference to the current annotation before rendering a task. Otherwise, we're seeing bugs where a text task renders the wrong annotation, when the task has multiple annotations (eg. transcription lines.)

`task.annotation` is also used to generate `task.isComplete`, so we may see errors with required task status reflecting the most recent mark, not the selected mark.

Package:
lib-classifier

Closes #1951.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Review by srallen — COMMENTED** [review]
Is it possible to add a spec that checks for this type of bug?

**Review by shaunanoordin — CHANGES_REQUESTED** [review]
## PR Review

This PR attempts to fix #1951, in which the sub-task popup will always show the sub-task of the _most-recently created_ mark, not the _currently selected_ mark.

Testing for 1951 looks good (the ["Apple", "Banana"] to ["Avocado", "Banana"] test), but I've discovered a ⚠️ new bug where you can't proceed to the next step.

### Testing

Testing on localhost + macOS + Chrome on projects:
- [blicksams's Transcription Task Testing](https://localhost:8080/?project=11300&env=production) - features a text transcription task
- I Fancy Cats - features an ellipsis tool (with sub-task) and point tool (no sub-task)
- [darkeshard's Transformers workflow](https://localhost:8080/?env=staging&project=1651&workflow=2628) - features a Point Task (with sub-task) and a second Point Task (no sub-task)

"Avocado, Banana" Test Steps:
- On any of the drawing tasks with sub-tasks, make Mark 1 and give it a sub-task value of, say, "Apple"
- make Mark 2 and give it a sub-task value of "Banana"
- select Mark 1 (expectation: I see the value "Apple") and change it to "Avocado"
- Proceed to submit the Classification
- Expectation: submit a classification where, for that drawing task, a value of ["Avocado", "Banana"] is submitted

Test Results 1:
- ✅ for the Transcription Task Testing, selecting the first Mark shows "Apple" and I'm then able to change it to "Avocado"...
- ❌ ...but if I attempt to go to the NEXT step, the classifier crashes with the error message: `Cannot read property 'update' of undefined`
- The error can be traced to...

```
src/plugins/tasks/DrawingTask/models/DrawingTask.js

function complete () {
  self.annotation.update(self.marks)
}
```

Test Results 2:
- ❌ In Transcription Task Testing, if I immediately click on the Next button without making any annotations, I receive the same message.

Test Results 3:
- ❌ On Transformers and I Fancy Cats, I'm unable to get past the first step. For these workflows, the first steps are Single Answer Questions, not even drawing tasks.
 
Screen Shot 2020-12-18 at 22 23 40

### Status

I'm not sure what's going on here... is the re-addition of task.setAnnotation somehow creating blank/incomplete annotations if the, uh, Tasks aren't fully loaded or something?

In any case, this may be moot since I'm testing #1953 in parallel and that looks _way_ better. I'm not sure if 1953 essentially makes 1952 redundant, but at the moment it seems that way. Please see review in 1953 for further deets.

**Comment by shaunanoordin** [comment]
Update on the #1953 review: looks like I ran into another issue that I need to dig deeper into. But that's a 2021 problem for me! I'll get back to investigating in January, but for now, good night!

File: packages/lib-classifier/src/components/Classifier/components/TaskArea/components/Tasks/components/Task/Task.js [hunk 1]
```diff
@@ -21,6 +21,7 @@ function useStores(task, stores) {
   if (!annotation) {
     annotation = classification.addAnnotation(task)
   }
+  task.setAnnotation(task)
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/TaskArea/components/Tasks/components/Task/Task.js:24]:
```suggestion
  task.setAnnotation(annotation)
```

**Comment by eatyourgreens** [comment]
@srallen it seems like it should be possible to test the task and annotation that are passed down as props. I had a quick look at the `SubTaskPopup` tests and those aren’t being tested at the moment.

**Comment by eatyourgreens** [comment]
#1953 has tests, and might be a better fix for the text task value bug.

**Comment by eatyourgreens** [comment]
@shaunanoordin there was a typo in the code. I wrote `task.setAnnotation(task)` instead of `task.setAnnotation(annotation)`. Should be fixed now. 🤦

**Comment by eatyourgreens** [comment]
`task.annotation` is used to generate `task.isComplete`. I haven't checked a transcription project, but I think this PR might fix an issue where `task.isComplete` reflects the last mark that you made, rather than the selected mark. Similar to the bug where the task value reflected the most recent mark, but not quite the same.

That should be testable too, so I'll circle back to this and write some tests. I'm setting this to draft until it's tested. Feel free to bump it if it's blocking the transcription task beta test.

**Review by shaunanoordin — APPROVED** [review]
## PR Review (Update)

I know this PR is still in draft mode (waiting on tests, me thinks) but I just wanted to give this a 👍 for what's done so far.

- The latest update (changing `task.setAnnotation(task)` to `task.setAnnotation(annotation)` fixed the "crashing when proceeding to the Single Text Task" that was present earlier in this branch.
- Testing on the Transcription Test, I Fancy Cats, and Transformers projects all looks good. I'm no longer seeing the "popup is showing last created mark, not selected mark" issue any longer (i.e. Apple, Banana test)
  - (Although I'll note that specific issue seemed solved by the parallel work in #1953)

> I haven't checked a transcription project, but I think this PR might fix an issue where task.isComplete reflects the last mark that you made, rather than the selected mark. Similar to the bug where the task value reflected the most recent mark, but not quite the same.

I'm not sure how this hypothetical bug would trigger, so I can't test for it - but I haven't encountered it either. So, no complaints from me at the moment.

### Status

LGTM

**Comment by eatyourgreens** [comment]
It turns out transcription lines can’t have incomplete annotations, so that bug would only affect drawing tasks.

**Comment by eatyourgreens** [comment]
I've set up a drawing task with a required question on each small point mark: https://frontend.preview.zooniverse.org/projects/eatyourgreens/-project-testing-ground/classify/workflow/3370?env=staging

It's impossible to save a task with an incomplete, required annotation. I think that means that the bug outlined here can't be triggered, and we can safely assume that `task.isComplete` (and `mark.isComplete`) will always be true for marks with required tasks.
Screenshot of the drawing task with a required question. A dialog prompts you to answer the question or delete your drawn mark.

I'm going to close this but feel free to re-open it if tasks and annotations being out of sync does turn out to cause problems.

**Comment by eatyourgreens** [comment]
I spoke too soon. Creating an incomplete mark makes all your marks incomplete.

1. Make a mark
2. Answer the question and save.
3. Make a new mark.
4. Close without saving.
5. Open your previous mark. It's now incomplete. Closing the dialog will delete it.

The reason for this is that each mark now has a reference to the incomplete annotation, which was deleted?

Screenshot of a drawing task with a required question. The question has been answered but you are prompted to close without saving, as if the answer doesn't exist.

**Comment by eatyourgreens** [comment]
Maybe `task.isComplete` should be `task.isComplete(annotation)` to make the dependence on an annotation explicit.

**Comment by eatyourgreens** [comment]
Closing this in favour of #1985.

**Pull request closed without merging**